### PR TITLE
UCT: Introduce iface_attr::max_num_eps configurable parameter

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Introduce `iface_attr::max_num_eps` configurable parameter

## Why ?

This is needed to implement UCP scoring for fallback on more salable transports when running on large scale

## How ?

1. Introduce new config value `MAX_NUM_EPS`
2. API change in `uct_iface_attr_t` - new field `max_num_eps`
3. By default this is `INF`, RC overrides it to `256`

### CODE REVIEW

File: src/uct/ib/cm/cm_iface.c [hunk 1]
```diff
@@ -421,6 +421,7 @@ static ucs_status_t uct_cm_iface_query(uct_iface_h tl_iface,
                                         UCT_IFACE_FLAG_PENDING  |
                                         UCT_IFACE_FLAG_CB_ASYNC |
                                         UCT_IFACE_FLAG_CONNECT_TO_IFACE;
+
```
Comment (brminich) [inline src/uct/ib/cm/cm_iface.c]: unrelated

**Comment by dmitrygx** [comment]
depends on #3967

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/11574/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>FAILED</b> on 8 of 26 workers (click for details)</summary>

**Note:** the logs will be deleted after 05-Aug-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/753dd6d23d0263169c284a4c0583c0f7d87acd9b/_main) | :question: ABORTED
[hpc-test-althca_W1](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/b4559f1185b46fc1cffe31fa541955d68c1e7960/hpc-test-althca_W1) | :question: ABORTED
[hpc-test-althca_W3](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/7a8e77318eb005ade4c0a4480b08f4e206fd859f/hpc-test-althca_W3) | :question: ABORTED
[hpc-test-althca_W0](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/72c200c80e8351c9a105fda4cdac88c585d3a0da/hpc-test-althca_W0) | :x: FAILURE
[hpc-test-althca_W2](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/a760b46cae53d9f1937b52e10c1e6fb342d99b9a/hpc-test-althca_W2) | :x: FAILURE
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/227d492d5ddefb2a2a66c4378e85a26659594a91/hpc-test-node-gpu_W0) | :x: FAILURE
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/d0cabf1e208b2eeb092caa9d68f32fbfb430087a/hpc-test-node-gpu_W2) | :x: FAILURE
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/4ba3732a87a0298429ca8fedb1399227d961b9f2/r-vmb-ppc-jenkins_W0) | :x: FAILURE
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/f1c55f92a70c87c01d5b2905ba55b385a457d6a6/r-vmb-ppc-jenkins_W1) | :x: FAILURE
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/8c340d14cb5f606a2fb7bef1fede01442ce11bd7/r-vmb-ppc-jenkins_W2) | :x: FAILURE
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/529915b5163e59dbe95c443651afccf35ffcbd5e/r-vmb-ppc-jenkins_W3) | :x: FAILURE
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/2f8b6568c52b81cd0eede46b5e72aa2e17e12e95/hpc-arm-cavium-jenkins_W0) | :question: UNKNOWN
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/b4803139c7857e0adf20e7860c0a06f4e4619aed/hpc-arm-cavium-jenkins_W1) | :question: UNKNOWN
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/2d1b8663b7eae45293028f8e6e2d1c495adc18db/hpc-arm-cavium-jenkins_W2) | :question: UNKNOWN
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/89e9afeb6749b6ea54f9eefccd40e7f8c77e7cbb/hpc-arm-cavium-jenkins_W3) | :question: UNKNOWN
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/ca9ce6cb9a404d76f04694284e2380310cd0d22f/hpc-arm-hwi-jenkins_W0) | :question: UNKNOWN
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/b47acaaf2d3392a3864f55a4bb65671802a1ac11/hpc-arm-hwi-jenkins_W1) | :question: UNKNOWN
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/418ca3e5fe970914e7f8a58f3701d2505c939439/hpc-arm-hwi-jenkins_W2) | :question: UNKNOWN
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/b4eed58c3d1b7aaa910f5a04c8b34556f0c409c7/hpc-test-node-gpu_W1) | :question: UNKNOWN
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/65f8e70addf27fb99e375ea8e4aab4fbc72fc9b6/hpc-test-node-gpu_W3) | :question: UNKNOWN
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/1a52434c6e88c894982034f230965804573530b1/hpc-test-node-legacy_W0) | :question: UNKNOWN
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/c3c2f564d80a69846f8f0e4cfc51341aab4a9481/hpc-test-node-legacy_W2) | :question: UNKNOWN
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/cc7139c47db0b9ec169b0c6127134d243ac88e6c/hpc-test-node-new_W0) | :question: UNKNOWN
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/8d6ba05c753012db3ac014a2f7c8e6339a47f8d6/hpc-test-node-new_W1) | :question: UNKNOWN
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/67bab2ade3f7452f01765fc3374a7a0c03b9a653/hpc-test-node-new_W2) | :question: UNKNOWN
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/6adc8bcf997528d40c3f17138e94d752/raw/6430f758c6be433c19af13e3e54c4e44ae86ff60/hpc-test-node-new_W3) | :question: UNKNOWN

File: src/uct/base/uct_iface.c [hunk 2]
```diff
@@ -564,5 +565,9 @@ ucs_config_field_t uct_iface_config_table[] = {
    "Level of network failure reporting",
    ucs_offsetof(uct_iface_config_t, failure), UCS_CONFIG_TYPE_ENUM(ucs_log_level_names)},
 
+  {"MAX_NUM_EPS", "inf",
+   "Maximum number of enpoints that UCT is able to create",
```
Comment (yosefe) [inline src/uct/base/uct_iface.c]: Maximum number of endpoints that the transport interface is able to create

File: src/uct/api/uct.h [hunk 3]
```diff
@@ -886,6 +886,7 @@ struct uct_iface_attr {
     double                   bandwidth;    /**< Maximal bandwidth, bytes/second */
     uct_linear_growth_t      latency;      /**< Latency model */
     uint8_t                  priority;     /**< Priority of device */
+    size_t                   max_num_eps;  /**< Maximum number of endpoints */
```
Comment (yosefe) [inline src/uct/api/uct.h]: maybe rename to max_eps?

File: src/uct/ugni/udt/ugni_udt_iface.c [hunk 4]
```diff
@@ -189,6 +189,8 @@ static ucs_status_t uct_ugni_udt_iface_query(uct_iface_h tl_iface, uct_iface_att
     iface_attr->latency.growth         = 0;
     iface_attr->bandwidth              = pow(1024, 2); /* bytes */
     iface_attr->priority               = 0;
+    iface_attr->max_num_eps            = iface->super.super.config.max_num_eps;
```
Comment (yosefe) [inline src/uct/ugni/udt/ugni_udt_iface.c]: maybe add function uct_base_iface_query()?

Comment (dmitrygx) [reply src/uct/ugni/udt/ugni_udt_iface.c]: yep, this was my idea - but it could do only:
```
memset(iface_attr, 0, sizeof(*iface_attr));
iface_attr->max_num_eps = iface->config.max_eps;
```

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/11628/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 29 workers (click for details)</summary>

**Note:** the logs will be deleted after 07-Aug-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/e9788af7bcf5817441020db98c4f6eee1a442dfb/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/509efc525c9a4fdaf0081ee5835b421350d2bd91/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/a1162c0656b6efd8814e327945c126c02ec98098/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/7f48204fd4be6e0ebdabc8bde0f9d1510c4e4ae2/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/4fd3417d9201bd9c5a62e4a8b52d9fc24fa7cdc4/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/38cb1f50397926347854d487457eb05934c58f4d/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/20293db9a1f01569dfdcecc75716db7241c044b6/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/730d66eaa7c081335fdbe45f399a868586ddcf7d/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/8c9643081927a87acf424a8abee19d53ee5406d7/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W0](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/a2d4fb584d60a8f474a575f9644cf7d27b78e21a/hpc-test-althca_W0) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W1](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/d939bf18d26e51e97c20712a0cb43abbcc10ca81/hpc-test-althca_W1) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W2](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/b74cd424fe34981d8ce044a7da681434b3f83a7d/hpc-test-althca_W2) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W3](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/fd5e1e1b5fcb7add53fd3b3eaffa92d476e0fde1/hpc-test-althca_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/0a9a7634d4a69df2c98d21018a6b5923ea245f7a/hpc-test-node-gpu_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/2dfa17aa6b05522f4a6f503bdc24fa8f093c77b5/hpc-test-node-gpu_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/27d22144c13b2adc5128271b8fc355c615f1482e/hpc-test-node-gpu_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/8ffd9c85d11ab4590ca47c44acdfc78d9ae8a0c4/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/cb2b23ee7bd8157806fc2ea74072238ab9441bfc/hpc-test-node-legacy_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/78525aff4f67ca7fac13c485fa52949e5649b2db/hpc-test-node-legacy_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/191ef909f4bf702a2db138e6157ea010848d736b/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/7e8e1ea7cc7e5c0ffbc0b8292c81ea79324f927f/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/60cd9870c75a9fa701920bd246841df6c45f283a/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/5bf76cb7d354bc8f7fbb9d668bb2607380c4ae7c/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/88df78d13f62937c2854433d5f6e229ab98934b1/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/93a60e92deb48652981f325c54f55fb130d7ef9f/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/6ee13640e69a79e9701f14050ad00b0c650f86e8/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/372f554e945b564fa1adf68184c29ee7a12455a3/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/d25cb0098f5ddb21565d97bd58e871e28f846bd8/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/afcbe76fa6761a5804fa8edcd35d1b3d/raw/771808980da22954f1eaa3cea0d8f94b0d9decf6/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

**Comment by yosefe** [comment]
@shamisp @tonycurtis can you pls take a look?

**Comment by tonycurtis** [comment]
> On Jul 31, 2019, at 7:41 AM, Yossi Itigin <notifications@github.com> wrote:
> 
> @shamisp <https://github.com/shamisp> @tonycurtis <https://github.com/tonycurtis> can you pls take a look?
> 

Comments etc. look good

T

**Comment by shamisp** [comment]
👍
